Closes #125 xportr 0.4.0 blog (#126)
pharmaverse/blog: posts/2024-03-18_xportr_0_4_0/xportr_0_4_0.qmd

---
title: "xportr 0.4.0"
author:
  - name: Sadchla Mascary
description: "Let's Talk About xportr 0.4.0, An R Package Within Pharmaverse Universe! "
date: "2024-03-18"
# please do not use any non-default categories.
# You can find the default categories in the repository README.md
categories: [ADaMs, xportr]
# feel free to change the image
image: "xportr.png"

---

<!--------------- typical setup ----------------->

```{r setup, include=FALSE}
long_slug <- "2024-03-18_xportr_0_4_0"
# renv::use(lockfile = "renv.lock")
```

<!--------------- post begins here ----------------->

# `XPT` Meets [`xportr`](https://atorus-research.github.io/xportr/)!

> In the pharmaceuticals and healthcare industries, it is crucial to maintain a standard structure for data exchange and regulatory submissions, enter `xpt` datasets! `XPT` datasets are binary files that are typically created by SAS software, they contain structured data, including variables, labels, and metadata. In order to develop `xpt` formatted files in R, let's introducing you to [`xportr`](https://atorus-research.github.io/xportr/).

Reviewer: Why is the second `XPT` all capital?

Reviewer: I think the last sentence should be rephrased, e.g.

```suggestion
In the pharmaceuticals and healthcare industries, it is crucial to maintain a standard structure for data exchange and regulatory submissions, enter `xpt` datasets! `XPT` datasets are binary files that are typically created by SAS software, they contain structured data, including variables, labels, and metadata. The  [`xportr`](https://atorus-research.github.io/xportr/) package was created specifically to develop `xpt` formatted files in R.
```

> [`xportr`](https://atorus-research.github.io/xportr/) is tools to build CDISC compliant data sets and check for CDISC compliance. By using  [`xportr`](https://atorus-research.github.io/xportr/), programmers are able to build these CDISC compliant files. 

Reviewer: ```suggestion
[`xportr`](https://atorus-research.github.io/xportr/) is a tool to build CDISC compliant data sets and check for CDISC compliance. By using  [`xportr`](https://atorus-research.github.io/xportr/), programmers are able to build these CDISC compliant files. 
```

# What is [`xportr`](https://atorus-research.github.io/xportr/)?

> [`xportr`](https://atorus-research.github.io/xportr/) has been developed through collaborations between developers from different parts of the industry [click here for a full list](https://atorus-research.github.io/xportr/authors.html). This collaborative effort ensures that [`xportr`](https://atorus-research.github.io/xportr/) meets the diverse needs and requirements of users across various domains. By leveraging insights and expertise from different stakeholders, [`xportr`](https://atorus-research.github.io/xportr/) continues to evolve and improve, maintaining its relevance and effectiveness in the ever-changing landscape of data management and analysis. With its meticulous design flow, users can expects to develop `xpt` files that has been vetted by many checks in the backend and CDISC compliant. 

Reviewer: ```suggestion
[`xportr`](https://atorus-research.github.io/xportr/) has been developed through collaborations between developers from different parts of the industry [click here for a full list](https://atorus-research.github.io/xportr/authors.html). This collaborative effort ensures that [`xportr`](https://atorus-research.github.io/xportr/) meets the diverse needs and requirements of users across various domains. By leveraging insights and expertise from different stakeholders, [`xportr`](https://atorus-research.github.io/xportr/) continues to evolve and improve, maintaining its relevance and effectiveness in the ever-changing landscape of data management and analysis. With its meticulous design flow, users can expect to develop `xpt` files that have been vetted by many checks in the backend and which are CDISC compliant. 
```

![](xportr_design_flow.png){fig-align="center" width="500"}

> [`xportr`](https://atorus-research.github.io/xportr/) is currently on version 0.4.0 and you can keep up with all of its improvements by follow the [`Changelog here`](https://atorus-research.github.io/xportr/news/index.html)

Reviewer: ```suggestion
[`xportr`](https://atorus-research.github.io/xportr/) is currently on version 0.4.0 and you can keep up with all of its improvements by following the [`Changelog here`](https://atorus-research.github.io/xportr/news/index.html)
```

# Why [`xportr`](https://atorus-research.github.io/xportr/)?

> Have I already mentioned that [`xportr`](https://atorus-research.github.io/xportr/) is an essential R package for generating `xpt` datasets! This package have functions to apply metadata information to your CDISC compliant datasets that were created in R, perform checks and provide useful feedbacks. In this blog post, we'll explore some high level of the capabilities of [`xportr`](https://atorus-research.github.io/xportr/), and its development.

Reviewer: ```suggestion
Have I already mentioned that [`xportr`](https://atorus-research.github.io/xportr/) is an essential R package for generating `xpt` datasets? This package includes functions to apply metadata information to your CDISC compliant datasets that were created in R, perform checks and provide useful feedbacks. In this blog post, we'll explore some high level  capabilities of [`xportr`](https://atorus-research.github.io/xportr/), and its development.
```

There are data deliverables and supporting documentation needed for a successful submission to Health Authorities, while you can find a walk through on how you can use `xportr` [`here`](https://atorus-research.github.io/xportr/articles/deepdive.html#what-goes-in-a-submission-to-a-health-authority), let's look at a high level example:

```{r,  message = FALSE, echo = FALSE }
library(xportr)
library(magrittr)
```

Looking at an `adsl` and the define metadata information
```{r,  message = FALSE, warnings = FALSE }
adsl <- data.frame(
  Subj = as.character(123, 456, 789),
  Different = c("a", "b", "c"),
  Val = c("1", "2", "3"),
  Param = c("param1", "param2", "param3")
)
```

and the define metadata information
```{r,  message = FALSE, warnings = FALSE}
metadata <- data.frame(
  dataset = "test",
  variable = c("Subj", "Param", "Val", "NotUsed"),
  type = c("numeric", "character", "numeric", "character"),
  format = NA,
  order = c(1, 3, 4, 2)
)
```

> Users can use the [`xportr_metadata()`](https://atorus-research.github.io/xportr/reference/xportr_metadata.html) function along with [`xportr_type()`](https://atorus-research.github.io/xportr/reference/xportr_type.html) and [`xportr_order()`](https://atorus-research.github.io/xportr/reference/xportr_order.html) to not only apply metadata attributes to your data, receive feedbacks on your data based on your spec, all while also performing valuable checks.

Reviewer: ```suggestion
Users can use the [`xportr_metadata()`](https://atorus-research.github.io/xportr/reference/xportr_metadata.html) function along with [`xportr_type()`](https://atorus-research.github.io/xportr/reference/xportr_type.html) and [`xportr_order()`](https://atorus-research.github.io/xportr/reference/xportr_order.html) to not only apply metadata attributes to your data but also receive feedback on your data based on your spec – all while also performing valuable checks.
```
What do you mean by "valuable checks"? Could you be more specific?

```{r, warnings = FALSE}
adsl %>%
  xportr_metadata(metadata, "test") %>%
  xportr_type() %>%
  xportr_order()
```

You can also use additional functions like [`xportr_write()`](https://atorus-research.github.io/xportr/reference/xportr_write.html) to generate your `xpt` file.

# Key Features of [`xportr`](https://atorus-research.github.io/xportr/):

- **Efficient Handling**: [`xportr`](https://atorus-research.github.io/xportr/) offers efficient avenues to produce `xpt` files, allowing users to manipulate clinical datasets with ease and export in `xpt` format in R.
- **Data Integrity**: [`xportr`](https://atorus-research.github.io/xportr/) preserves data integrity by retaining variable labels, formats, and other metadata during data export operations.
- **User-Friendly Interface**: [`xportr`](https://atorus-research.github.io/xportr/) provides a user-friendly interface with intuitive functions for generating `xpt` datasets and performing data manipulation tasks.

> # About the other kid on the blog, (`JSON`)! Is The Future `XPT` or JSON?

Reviewer 2: Haha! Amazing!  Can you make a reference to the datasetjson blog post on Atorus' site and the datasetjson package?  

@mstackhouse maybe a follow up blog post on pharamverse is needed about datasetjson??

`XPT` datasets are standardized, structured files used for storing clinical trial data, making them essential for regulatory submissions and data analysis. They remain indispensable in certain industries and with advancements in data management and analysis tools, the future of `xpt` datasets lies in improved efficiency, compatibility, and ensuring seamless integration with emerging technologies and data formats.

> As data continues to play a central role in decision-making and innovation, the future of JavaScript Object Notation (JSON) datasets are worth exploring. JSON datasets offer flexibility, simplicity, and interoperability, making them increasingly popular for data storage and exchange. JSON's lightweight and human-readable format make it suitable for a wide range of applications, including web development, APIs, IoT, and data storage. While `xpt` datasets remain prevalent in certain industries, there's a growing interest in JSON datasets across various domains. Although we will not be exploring JSON datasets, this formatted data type was worth mentioning in this blog.

Reviewer: ```suggestion
As data continues to play a central role in decision-making and innovation, the future of JavaScript Object Notation (JSON) datasets is worth exploring. JSON datasets offer flexibility, simplicity, and interoperability, making them increasingly popular for data storage and exchange. JSON's lightweight and human-readable format make it suitable for a wide range of applications, including web development, APIs, IoT, and data storage. While `xpt` datasets remain prevalent in certain industries, there's a growing interest in JSON datasets across various domains. Although we will not be exploring JSON datasets, this formatted data type was worth mentioning in this blog.
```

I have heard about JSON being discussed as a standard data format in the industry before. Would you be able to link here some further reading on this?

# Let's Recap!

> This blog post highlighted the benefits of using the [`xportr`](https://atorus-research.github.io/xportr/) package, showcased an example of [`xportr`](https://atorus-research.github.io/xportr/) function, and discussed the growing interest of JSON datasets in the data landscape. We've also discussed [`xportr`](https://atorus-research.github.io/xportr/)'s development through collaborations between developers from different parts of the industry, highlighting its versatility and relevance in various domains. Which gives me a nice transition to my call to action!! If you haven't used [`xportr`](https://atorus-research.github.io/xportr/) before, then consider this blog as your official invite to test it! Reap the benefit of collaborative efforts across industries and the pharma world. If are already a user of [`xportr`](https://atorus-research.github.io/xportr/) or you are experiencing some issues with a particular functions, then chances are someone else could either be experiencing a similar issue or have an answer so reach out to the team by creating an issue [`here`](https://github.com/atorus-research/xportr/issues).  

Reviewer: ```suggestion
This blog post highlighted the benefits of using the [`xportr`](https://atorus-research.github.io/xportr/) package, showcased an example of [`xportr`](https://atorus-research.github.io/xportr/) function, and discussed the growing interest of JSON datasets in the data landscape. We've also discussed [`xportr`](https://atorus-research.github.io/xportr/)'s development through collaborations between developers from different parts of the industry, highlighting its versatility and relevance in various domains. Which gives me a nice transition to my call to action!! If you haven't used [`xportr`](https://atorus-research.github.io/xportr/) before, then consider this blog as your official invite to test it! Reap the benefit of collaborative efforts across industries and the pharma world. If are already a user of [`xportr`](https://atorus-research.github.io/xportr/) or you are experiencing some issues with a particular functions, then chances are someone else could either be experiencing a similar issue or has an answer to it. So reach out to the team by creating an issue [`here`](https://github.com/atorus-research/xportr/issues).  
```

```{r,  message = FALSE, echo = FALSE }
library(xportr)
library(magrittr)
```

Looking at an `adsl` and the define metadata information
```{r,  message = FALSE, warnings = FALSE }
adsl <- data.frame(
  Subj = as.character(123, 456, 789),
  Different = c("a", "b", "c"),
  Val = c("1", "2", "3"),
  Param = c("param1", "param2", "param3")
)
```

and the define metadata information
```{r,  message = FALSE, warnings = FALSE}
metadata <- data.frame(
  dataset = "test",
  variable = c("Subj", "Param", "Val", "NotUsed"),
  type = c("numeric", "character", "numeric", "character"),
  format = NA,
  order = c(1, 3, 4, 2)
)
```

Users can use the [`xportr_metadata()`](https://atorus-research.github.io/xportr/reference/xportr_metadata.html) function along with [`xportr_type()`](https://atorus-research.github.io/xportr/reference/xportr_type.html) and [`xportr_order()`](https://atorus-research.github.io/xportr/reference/xportr_order.html) to not only apply metadata attributes to your data but also receive feedback on your data based on your spec. Additionally, some of those functions performing valuable checks like:

>   * Variable names must start with a letter (not an underscore), be comprised of only uppercase letters (A-Z), numerals (0-9) and be free of non-ASCII characters, symbols, and underscores.

Reviewer 2: Could we list out the other functions here.  The xportr_label, xportr_format and xportr_write with the strict checks should get a mention here for sure.

Reviewer 2: oh and xportr_length
